### PR TITLE
[asr->fortran] add more visitors and enable tests

### PR DESCRIPTION
**Pull request by khushi-411**
#### Scope of the PR
- Adds complex type support in the fortran backend.
    - add support for ComplexConstructor, ComplexConstant, ComplexRe, and ComplexIm visitors
- Support for IntegerBitLen
- Intrinsics functions: Max, Min, Any, Sqrt
- String visitor: repeat
- Ichar and Iachar
#### TODO
- [x] add/enable some more tests
~- [ ] implement FunctionCall visitor in another PR~
- [x] check for edge cases

### CODE REVIEW

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
Enable tests for `repeat(..)`, `complex`, `sqrt` and `integerbinlen` as well.
For every implementation, we either add a new test or enable pre-existing tests, so that we always test the implementation using our CI.

**Comment by khushi-411** [comment]
Yeah, I haven't yet included those because now the remaining tests require a `FunctionCall` visitor; I'll add that in a different PR. And then update this PR as needed. Thanks!

File: integration_tests/bits_06.f90 [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+integer(1) :: int1 = 1
+integer(2) :: int2 = 4
+integer(4) :: int4 = 41
+integer(8) :: int8 = 411
+
+print *, bit_size(int1)
+print *, bit_size(int2)
+print *, bit_size(int4)
+print *, bit_size(int8)
+
```
Comment (Shaikh-Ubaid) [inline integration_tests/bits_06.f90:13]: Please add asserts to the test so that we ensure that the tests run correctly.

Comment (khushi-411) [reply integration_tests/bits_06.f90:27]: Done!

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 2]
```diff
@@ -956,7 +974,13 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
     }
 
-    // void visit_StringRepeat(const ASR::StringRepeat_t &x) {}
+    void visit_StringRepeat(const ASR::StringRepeat_t &x) {
+        visit_expr(*x.m_left);
+        std::string str = s;
+        visit_expr(*x.m_right);
+        std::string n = s;
+        s = "repeat(" + str + ", " + n + ")";
+    }
```
Comment (khushi-411) [inline src/libasr/codegen/asr_to_fortran.cpp:983]: @Shaikh-Ubaid, this PR is ready, except for this part. It currently generates: (for intrinsics_53)
```f90
module lfortran_intrinsic_iso_fortran_env
end module lfortran_intrinsic_iso_fortran_env
module lfortran_intrinsic_builtin
end module lfortran_intrinsic_builtin
module lfortran_intrinsic_string
end module lfortran_intrinsic_string
program intrinsics_53
implicit none
integer(4), save :: n4 = 3
integer(8), save :: n8 = 3
print *, repeat@repeati32("a", n4)

print *, repeat@repeati64("a", n8)

if (repeat@repeati32("a", n4)
 /= "aaa") then
    error stop
end if
if (repeat@repeati64("a", n8)
 /= "aaa") then
    error stop
end if
end program intrinsics_53
```
I believe the code generation responsible for `repeat@repeati32` is unrelated to this part. I was just looking for the reason. Do you have any suggestions on it? Thanks!

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_fortran.cpp:983]: I think we need to visit the symbols present in a module. 

https://github.com/lfortran/lfortran/blob/34e92d66f0323f0792dc1d6ae8c924d3442117eb/src/libasr/codegen/asr_to_fortran.cpp#L253-L265

Currently, we skip visiting those symbols and therefore in the fortran code, we did not get the definition of "repeati32/i64" generated.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_fortran.cpp:983]: >I think we need to visit the symbols present in a module.

I would suggest to work on this in a separate PR.

>I believe the code generation responsible for repeat@repeati32 is unrelated to this part. I was just looking for the reason. Do you have any suggestions on it? Thanks!

Also, yes the code you implemented is unrelated to `visit_Module()`.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 3]
```diff
@@ -698,7 +703,13 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
     /********************************** Expr **********************************/
     // void visit_IfExp(const ASR::IfExp_t &x) {}
 
-    // void visit_ComplexConstructor(const ASR::ComplexConstructor_t &x) {}
+    void visit_ComplexConstructor(const ASR::ComplexConstructor_t &x) {
+        visit_expr(*x.m_re);
+        std::string __re_ = s;
+        visit_expr(*x.m_im);
+        std::string __im_ = s;
+        s = "(" + __re_ + ", " + __im_ + ")";
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_fortran.cpp:756]: I am curious to know why we use names `__re_` and `__im_` here and also in `visit_ComplexConstant`? Did we use these names anywhere previously before this PR? 

Do simple names like `re` and `im` conflict with any predefined keyword?

File: integration_tests/bits_06.f90 [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+size4 = bit_size(int4)
+if (size4 /= 32) error stop
+
+size8 = bit_size(int8)
+if (size8 /= 64) error stop
+
+print *, bit_size(int1)
+print *, bit_size(int2)
+print *, bit_size(int4)
+print *, bit_size(int8)
```
Comment (Shaikh-Ubaid) [inline integration_tests/bits_06.f90:26]: Please print the values before the assert. So that way if the assert failed, there would atleast be a value printed on the console that could be helpful to debug.

Also, I think you can simply reuse the previous return values of `bit_size()` while printing. We need not recompute the result while printing. 

For example you can have:
```fortran
size1 = bit_size(int1)
print *, size1
if (size1 /= 8) error stop
```

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 5]
```diff
@@ -1312,29 +1336,57 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
+        switch (arg_kind) {
+            case 1: s = "_lpython_bit_length1(" + s + ")"; break;
+            case 2: s = "_lpython_bit_length2(" + s + ")"; break;
+            case 4: s = "_lpython_bit_length4(" + s + ")"; break;
+            case 8: s = "_lpython_bit_length8(" + s + ")"; break;
+            default: {
+                throw LCompilersException("Unsupported Integer Kind: "
+                    + std::to_string(arg_kind));
+            }
+        }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_fortran.cpp:1367]: I think this can be simplified as:
```suggestion
        s = "bit_size(" + s + ")";
```

**Comment by Shaikh-Ubaid** [comment]
@khushi-411 For each test case that you enabled, did you verify that the fortran backend generated output closely matches or is exactly similar to the test source code?

**Review by Shaikh-Ubaid — COMMENTED** [review]
I shared some comments above. Overall the PR looks good to me. Thanks for the contributions!

**Comment by Shaikh-Ubaid** [comment]
Are there any visitors that are implemented in this PR but not/yet-to-be tested? I would suggest to create an issue about it so that we don't forget.

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 2 above)
Comment (khushi-411) [reply src/libasr/codegen/asr_to_fortran.cpp]: I see. Thanks for describing and confirming it, @Shaikh-Ubaid!

I'll create an issue regarding the same.

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 3 above)
Comment (khushi-411) [reply src/libasr/codegen/asr_to_fortran.cpp]: Oops, sorry. There's no technical reason. I like naming using that, so I did it while working, but it slipped off my mind to undo those changes.

File: integration_tests/bits_06.f90 (hunk 4 above)
Comment (khushi-411) [reply integration_tests/bits_06.f90]: Nice nitpick. updated the PR. Thanks, Shaikh!

**Comment by khushi-411** [comment]
Thanks for reviewing, @Shaikh-Ubaid! Addressed all the reviews. I think this PR is ready to go now. :-)

**Comment by Shaikh-Ubaid** [comment]
Great, Thanks for addressing the reviews! I appreciate it. Please, could you respond to the following queries and then it would be good to merge:

- @khushi-411 For each test case that you enabled, did you verify that the fortran backend generated output closely matches or is exactly similar to the test source code?
- Are there any visitors that are implemented in this PR but not/yet-to-be tested? (You need not revert them. It is good to just know about them)

**Comment by khushi-411** [comment]
> @khushi-411 For each test case that you enabled, did you verify that the fortran backend generated output closely matches or is exactly similar to the test source code?

Yes, I have confirmed it; the test and generated codes are the same. Thanks!

> Are there any visitors that are implemented in this PR but not/yet-to-be tested? (You need not revert them. It is good to just know about them)

`repeat` is left because it currently does not support symbols in Module visitor. I created an issue #2737

One more thing: is there any reason why `visit_IntrinsicScalarFunction` accepts a single argument? Min and max tests are also left due to this. https://github.com/lfortran/lfortran/blob/5ecd01fa91482265072bcfb9197ec9326cb09a85/src/libasr/codegen/asr_to_fortran.cpp#L773-L790

**Comment by Shaikh-Ubaid** [comment]
>is there any reason why visit_IntrinsicScalarFunction accepts a single argument? Min and max tests are also left due to this.

I think there is no specific reason. For example in the `llvm` backend you can see we access several args inside `visit_IntrinsicScalarFunction` here:
https://github.com/lfortran/lfortran/blob/34e92d66f0323f0792dc1d6ae8c924d3442117eb/src/libasr/codegen/asr_to_llvm.cpp#L1923-L1937

To support `min`, `max` or other intrinsics, you might need to update the code appropriately, for example removing `LCOMPILERS_ASSERT(x.n_args == 1)` and other updates. I think they can be supported in a separate PR.

**Comment by khushi-411** [comment]
Got it! Thanks, @Shaikh-Ubaid!

**Comment by Shaikh-Ubaid** [comment]
>Yes, I have confirmed it; the test and generated codes are the same. Thanks!

Thanks for the confirmation.

>repeat is left because it currently does not support symbols in Module visitor. I created an issue https://github.com/lfortran/lfortran/issues/2737

Thanks for sharing!

**Review by Shaikh-Ubaid — APPROVED** [review]
Great work! It looks good to me. Thank you so much for this!

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 6]
```diff
@@ -764,6 +775,8 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
         switch (x.m_intrinsic_id) {
             SET_INTRINSIC_NAME(Abs, "abs");
             SET_INTRINSIC_NAME(Exp, "exp");
+            SET_INTRINSIC_NAME(Max, "max");
+            SET_INTRINSIC_NAME(Min, "min");
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_fortran.cpp:779]: I think `min` or `max` tests is not enabled, along with it let's fix the arguments as well.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 7]
```diff
@@ -130,6 +130,11 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
                 r += std::to_string(down_cast<ASR::Real_t>(t)->m_kind);
                 r += ")";
                 break;
+	    } case ASR::ttypeType::Complex: {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_fortran.cpp:133]:
```suggestion
        } case ASR::ttypeType::Complex: {
```

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 8]
```diff
@@ -785,6 +798,7 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
     void visit_IntrinsicArrayFunction(const ASR::IntrinsicArrayFunction_t &x) {
         std::string out;
         switch (x.m_arr_intrinsic_id) {
+            SET_ARR_INTRINSIC_NAME(Any, "any");
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_fortran.cpp:801]: `any` test is not enabled.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 9]
```diff
@@ -956,7 +974,13 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
     }
 
-    // void visit_StringRepeat(const ASR::StringRepeat_t &x) {}
+    void visit_StringRepeat(const ASR::StringRepeat_t &x) {
+        this->visit_expr(*x.m_left);
+        std::string str = s;
+        this->visit_expr(*x.m_right);
+        std::string n = s;
+        s = "repeat(" + str + ", " + n + ")";
+    }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_fortran.cpp:983]: Test is not enabled

**Review by Thirumalai-Shaktivel — APPROVED** [review]
Rest looks good, thank you @khushi-411!!

The following changes can be addressed using a new PR.

**Comment by khushi-411** [comment]
Thanks, @Thirumalai-Shaktivel, for reviewing the PR!

Sure, I'll submit another PR to add these tests and formatting changes.
